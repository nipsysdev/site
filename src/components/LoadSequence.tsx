'use client';

import { Typography } from '@acid-info/lsd-react/client/Typography';
import { useRouter } from 'next/navigation';
import { useEffect, useMemo, useState } from 'react';
import type { Lang } from '@/constants/lang';
import { routing } from '@/i18n/intl';

interface BootStep {
  message: string;
  delay: number;
}

export default function LoadSequence() {
  const [currentStep, setCurrentStep] = useState(0);
  const [detectedLocale, setDetectedLocale] = useState<string>('en');
  const router = useRouter();

  // Detect user locale
  useEffect(() => {
    const browserLocale = navigator.language.split('-')[0];
    const supportedLocale = routing.locales.includes(browserLocale as Lang)
      ? browserLocale
      : routing.defaultLocale;
    setDetectedLocale(supportedLocale);
  }, []);

  const steps: BootStep[] = useMemo(
    () => [
      { message: '> Connection initialized', delay: 200 },
      { message: '> Detecting user locale...', delay: 400 },
      { message: `> Selecting proper locale [${detectedLocale}]`, delay: 300 },
      { message: `> Redirecting to /${detectedLocale}`, delay: 500 },
    ],
    [detectedLocale],
  );

  useEffect(() => {
    if (currentStep < steps.length) {
      const timer = setTimeout(() => {
        setCurrentStep(currentStep + 1);
      }, steps[currentStep].delay);

      return () => clearTimeout(timer);
    } else {
      // All steps done, redirect
      router.push(`/${detectedLocale}`);
    }
  }, [currentStep, steps, detectedLocale, router]);

  return (
    <div className="p-8">
      <div className="flex flex-col">
        {steps.slice(0, currentStep).map((step, index) => (
          <Typography key={`step-${index}-${step.message}`} variant="subtitle2">
            {step.message}
          </Typography>
        ))}
        {currentStep < steps.length && (
          <div className="flex items-center">
            <span className="animate-pulse">â–ˆ</span>
          </div>
        )}
      </div>
    </div>
  );
}
