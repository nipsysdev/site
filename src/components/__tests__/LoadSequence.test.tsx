import { render, screen, waitFor } from '@testing-library/react';
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
import LoadSequence from '../LoadSequence';

// Mock Typography component
vi.mock('@acid-info/lsd-react/client/Typography', () => ({
  Typography: ({ children, ...props }: { children: React.ReactNode }) => (
    <div {...props}>{children}</div>
  ),
}));

// Mock next/navigation
const mockPush = vi.fn();
vi.mock('next/navigation', () => ({
  useRouter: () => ({
    push: mockPush,
  }),
}));

// Mock the i18n routing
vi.mock('@/i18n/intl', () => ({
  routing: {
    locales: ['en', 'fr'],
    defaultLocale: 'en',
  },
}));

// Store original navigator.language
const originalLanguage = navigator.language;

describe('LoadSequence', () => {
  beforeEach(() => {
    vi.clearAllMocks();
    // Reset to default English
    Object.defineProperty(navigator, 'language', {
      writable: true,
      value: 'en-US',
    });
  });

  afterEach(() => {
    // Restore original navigator.language
    Object.defineProperty(navigator, 'language', {
      writable: true,
      value: originalLanguage,
    });
  });

  it('should render the component with cursor', () => {
    render(<LoadSequence />);

    // Should show the blinking cursor initially
    const cursor = screen.getByText('â–ˆ');
    expect(cursor).toBeInTheDocument();
    expect(cursor).toHaveClass('animate-pulse');
  });

  it('should eventually show first step', async () => {
    render(<LoadSequence />);

    // Wait for the first step to appear
    await waitFor(
      () => {
        expect(
          screen.getByText('> Detecting user locale...'),
        ).toBeInTheDocument();
      },
      { timeout: 1000 },
    );
  });

  it('should eventually show locale detection step', async () => {
    render(<LoadSequence />);

    // Wait for the locale detection step
    await waitFor(
      () => {
        expect(
          screen.getByText('> Detecting user locale...'),
        ).toBeInTheDocument();
      },
      { timeout: 2000 },
    );
  });

  it('should detect English locale from browser by default', async () => {
    render(<LoadSequence />);

    // Wait for locale selection step
    await waitFor(
      () => {
        expect(
          screen.getByText('> Selecting proper locale [en]'),
        ).toBeInTheDocument();
      },
      { timeout: 2000 },
    );
  });

  it('should detect French locale when set', async () => {
    // Set navigator.language to French before rendering
    Object.defineProperty(navigator, 'language', {
      writable: true,
      value: 'fr-FR',
    });

    render(<LoadSequence />);

    // Wait for locale selection step with French
    await waitFor(
      () => {
        expect(
          screen.getByText('> Selecting proper locale [fr]'),
        ).toBeInTheDocument();
      },
      { timeout: 2000 },
    );
  });

  it('should fallback to English for unsupported language', async () => {
    // Set navigator.language to unsupported language
    Object.defineProperty(navigator, 'language', {
      writable: true,
      value: 'es-ES',
    });

    render(<LoadSequence />);

    // Should fallback to English
    await waitFor(
      () => {
        expect(
          screen.getByText('> Selecting proper locale [en]'),
        ).toBeInTheDocument();
      },
      { timeout: 2000 },
    );
  });

  it('should eventually redirect to detected locale', async () => {
    render(<LoadSequence />);

    // Wait for redirect to happen
    await waitFor(
      () => {
        expect(mockPush).toHaveBeenCalledWith('/en');
      },
      { timeout: 3000 },
    );
  });

  it('should redirect to French locale when detected', async () => {
    // Clear previous calls
    mockPush.mockClear();

    // Set navigator.language to French
    Object.defineProperty(navigator, 'language', {
      writable: true,
      value: 'fr-CA',
    });

    render(<LoadSequence />);

    // Wait for French redirect
    await waitFor(
      () => {
        expect(mockPush).toHaveBeenCalledWith('/fr');
      },
      { timeout: 3000 },
    );
  });

  it('should handle partial locale codes correctly', async () => {
    // Set navigator.language to a specific region code
    Object.defineProperty(navigator, 'language', {
      writable: true,
      value: 'en-GB',
    });

    render(<LoadSequence />);

    // Should still detect as English
    await waitFor(
      () => {
        expect(
          screen.getByText('> Selecting proper locale [en]'),
        ).toBeInTheDocument();
      },
      { timeout: 2000 },
    );
  });

  it('should match snapshot', () => {
    const { container } = render(<LoadSequence />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
