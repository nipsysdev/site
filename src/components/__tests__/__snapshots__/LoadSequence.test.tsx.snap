// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoadSequence > should match snapshot 1`] = `
<div
  class="p-8"
>
  <div
    class="flex flex-col"
  >
    <div
      class="flex items-center"
    >
      <span
        class="animate-pulse"
      >
        â–ˆ
      </span>
    </div>
  </div>
</div>
`;
